Jenkinsfile

```diff
@@ -100,97 +47,150 @@ node(windows) {
-		sh 'mkdir -p tmp'
-		sh 'git describe --all > tmp/GIT_BRANCH'
-		sh 'git rev-parse HEAD > tmp/GIT_COMMIT'
-		git_branch = readFile('tmp/GIT_BRANCH').trim()
-		git_commit = readFile('tmp/GIT_COMMIT').trim()
-
-		if (git_branch != "remotes/origin/master") {
-			// Determine actual PR commit, if necessary
-			sh 'git rev-parse HEAD | git log --pretty=%P -n 1 --date-order > tmp/MERGE_COMMIT_PARENTS'
-			sh 'cat tmp/MERGE_COMMIT_PARENTS'
-			merge_commit_parents = readFile('tmp/MERGE_COMMIT_PARENTS').trim()
-			if (merge_commit_parents.length() > 40) {
-				echo 'More than one merge commit parent signifies that the merge commit is not the PR commit'
-				echo "Changing git_commit from '${git_commit}' to '${merge_commit_parents.take(40)}'"
-				git_commit = merge_commit_parents.take(40)
-			} else {
-				echo 'Only one merge commit parent signifies that the merge commit is also the PR commit'
-				echo "Keeping git_commit as '${git_commit}'"
+stage 'Lint and test container'
+parallel(
+	lint: {
+		node(linux) {
+			sh "docker run --rm ${test_image} lint"
+		}
+	},
+	test: {
+		node(linux) {
+			withCredentials([[$class: 'StringBinding',
+												credentialsId: '995d99a7-466b-4beb-bf75-f3ba91cbbc18',
+												variable: 'CODECOV_TOKEN']]) {
+				sh "docker run -e CODECOV_TOKEN=\${CODECOV_TOKEN} --rm ${test_image} sh -c 'test-cover.sh && curl -s https://codecov.io/bash | bash'"
 			}
 		}
 	}
-}
+)
 
 stage 'Build and Upload Artifacts'
 
+def gcs_bucket = "gs://workflow-cli"
+
+def upload_artifacts = { boolean cache ->
+	headers  = "-h 'x-goog-meta-git-branch:${git_branch}' "
+	headers += "-h 'x-goog-meta-git-sha:${git_commit}' "
+	headers += "-h 'x-goog-meta-ci-job:${env.JOB_NAME}' "
+	headers += "-h 'x-goog-meta-ci-number:${env.BUILD_NUMBER}' "
+	headers += "-h 'x-goog-meta-ci-url:${env.BUILD_URL}'"
+	if(!cache) {
+		headers += ' -h "Cache-Control:no-cache"'
+	}
+
+	script = "sh -c 'echo \${GCS_KEY_JSON} | base64 -d - > /tmp/key.json "
+	script += "&& gcloud auth activate-service-account -q --key-file /tmp/key.json "
+	script += "&& gsutil -mq ${headers} cp -a public-read -r /upload/* ${gcs_bucket} "
+	script += "&& rm -rf /upload/*'"
+
+	return script
+}
+
+def mktmp = {
+	// Create tmp directory to store files
+	sh 'mktemp -d > tmp_dir'
+	tmp = readFile('tmp_dir').trim()
+	sh 'rm tmp_dir'
+	return tmp
+}
+
 parallel(
 	revision: {
 		node(linux) {
-			def gopath = gopath_linux()
-			def workdir = workdir_linux(gopath)
 
-			dir(workdir) {
-					checkout scm
-
-					if (git_branch != "remotes/origin/master") {
-						echo "Skipping build of 386 binaries to shorten CI for Pull Requests"
-						env.BUILD_ARCH = "amd64"
-					}
+			flags = ""
+			if (git_branch != "remotes/origin/master") {
+				echo "Skipping build of 386 binaries to shorten CI for Pull Requests"
+				flags += "-e BUILD_ARCH=amd64"
+			}
 
-					make 'bootstrap'
-					env.REVISION = git_commit.take(7)
-					make 'build-revision'
+			tmp_dir = mktmp()
```

Comment:
nit: maybe consider containing this logic (until the `rm -rf ${tmp_dir}`) into a method?  It looks like the only difference between here and if `git_branch == "remotes/origin/master"` is the cache bool?